File: app/src/main/java/in/completecourse/questionbank/fragment/HomeFragment.java
package in.completecourse.questionbank.fragment;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.ViewPager;

import com.google.android.ads.mediationtestsuite.MediationTestSuite;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.material.tabs.TabLayout;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;

import org.json.JSONException;
import org.json.JSONObject;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import cz.msebera.android.httpclient.HttpResponse;
import cz.msebera.android.httpclient.client.HttpClient;
import cz.msebera.android.httpclient.client.methods.HttpPost;
import cz.msebera.android.httpclient.entity.StringEntity;
import cz.msebera.android.httpclient.impl.client.DefaultHttpClient;
import cz.msebera.android.httpclient.params.HttpConnectionParams;
import cz.msebera.android.httpclient.util.EntityUtils;
import in.completecourse.questionbank.R;
import in.completecourse.questionbank.ScanActivity;
import in.completecourse.questionbank.SearchActivity;
import in.completecourse.questionbank.SubjectActivity;
import in.completecourse.questionbank.adapter.ImageAdapter;
import in.completecourse.questionbank.adapter.SliderAdapter;
import in.completecourse.questionbank.app.AppConfig;
import in.completecourse.questionbank.helper.HelperMethods;
import in.completecourse.questionbank.helper.HttpHandler;
import in.completecourse.questionbank.helper.PrefManager;
import in.completecourse.questionbank.model.CardModel;
import in.completecourse.questionbank.model.Update;

public class HomeFragment extends Fragment {

    private ArrayList<Update> updateList;
    private static String urlQR;
    private ProgressDialog pDialog;
    @BindView(R.id.viewPager)
    ViewPager mViewPager;
    @BindView(R.id.indicator)
    TabLayout indicator;
    @BindView(R.id.edTv_request_book)
    EditText editText;
    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;
    private PrefManager prefManager;
    private FirebaseFirestore db;

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data)  {
        if (requestCode == 100 && resultCode == 100 && data != null){
            urlQR = data.getStringExtra("url");
            new GetBookName(HomeFragment.this).execute();
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_home, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        db = FirebaseFirestore.getInstance();

        prefManager = new PrefManager(view.getContext().getApplicationContext());

        ArrayList<CardModel> cardList = new ArrayList<>();
        cardList.add(new CardModel(R.drawable.manual_search, "Manual Search"));
        cardList.add(new CardModel(R.drawable.scan_qr, "Scan QR Code"));
        recyclerView.setHasFixedSize(true);
        //use a linear layout manager
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getContext() ,2, RecyclerView.VERTICAL, false);
        recyclerView.setLayoutManager(gridLayoutManager);

        //specify an adapter
        ImageAdapter recyclerViewAdapter = new ImageAdapter(cardList, getContext());
        recyclerView.setAdapter(recyclerViewAdapter);
        recyclerView.addOnItemTouchListener(new ImageAdapter.RecyclerTouchListener(getContext(), position -> {
            if (position == 0) {
                Intent intent = new Intent(getContext(), SearchActivity.class);
                startActivity(intent);
            }else if(position == 1){
                Intent qrCodeActivityIntent = new Intent(getContext(), ScanActivity.class);
                startActivityForResult(qrCodeActivityIntent, 100);
            }
        }));

        if (isNetworkAvailable()){
            getUpdates();
        }else{
            Toast.makeText(view.getContext(), "Please check your internet connection.", Toast.LENGTH_SHORT).show();
        }

        MediationTestSuite.launch(getContext());
    }

    @OnClick(R.id.btn_request_now)
    void requestBook(){
        if (isNetworkAvailable()) {
            String id = HelperMethods.INSTANCE.generateChecksum();
            String userrequest = editText.getText().toString().trim();
            String mUserId = prefManager.getUserId();
            if (!userrequest.isEmpty()) {
                JSONObject dataObj = new JSONObject();
                try {
                    dataObj.putOpt("id", id);
                    dataObj.putOpt("uid", mUserId);
                    dataObj.putOpt("userrequest", userrequest);
                    JSONTransmitter jsonTransmitter = new JSONTransmitter(HomeFragment.this);
                    jsonTransmitter.execute(dataObj);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }else
                Toast.makeText(getContext(), "Please enter your request", Toast.LENGTH_SHORT).show();
        }else
            Toast.makeText(getContext(), "Please check your internet connection", Toast.LENGTH_SHORT).show();
    }

    private class SliderTimer extends TimerTask {
        @Override
        public void run() {
            if (getActivity() != null){
                getActivity().runOnUiThread(() -> {
                    if (mViewPager.getCurrentItem() < updateList.size() - 1) {
                        mViewPager.setCurrentItem(mViewPager.getCurrentItem() + 1);
                    } else {
                        mViewPager.setCurrentItem(0);
                    }
                });
            }
        }
    }

    /**
     * Checks if there is Internet accessible.
     * @return True if there is Internet. False if not.
     */
    private boolean isNetworkAvailable() {
        NetworkInfo activeNetworkInfo = null;
        if (getActivity() != null){
            ConnectivityManager connectivityManager = (ConnectivityManager) ((getActivity())).getSystemService(Context.CONNECTIVITY_SERVICE);
            if (connectivityManager != null){
                activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
            }
        }
        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
    }

    private static class GetBookName extends AsyncTask<Void, Void, Void> {
        private final WeakReference<HomeFragment> activityWeakReference;

        GetBookName(HomeFragment context){
            activityWeakReference = new WeakReference<>(context);
        }

        @Override
        protected Void doInBackground(Void... arg0) {
            final HomeFragment activity = activityWeakReference.get();
            HttpHandler sh = new HttpHandler();
            //tokenNumber = url.substring(url.lastIndexOf('/') + 1, url.length() - 4);
            //Log.e("token", tokenNumber);
            String url = urlQR + "/get";
            // Making a request to url and getting response
            String jsonStr = sh.makeServiceCall(url);
            //Log.e("Company Activity ",  "Response from url: " + jsonStr);

            if (jsonStr != null) {
                try {
                    JSONObject jsonObject = new JSONObject(jsonStr);
                    String studentSubject = jsonObject.getString("subjectkiid");
                    String studentClass = jsonObject.getString("uclass");
                    //Log.e("class", studentClass);
                    //Log.e("subject", studentSubject);
                    Intent intent = new Intent(activity.getActivity(), SubjectActivity.class);
                    intent.putExtra("uclass", studentClass);
                    intent.putExtra("subjectkiid", studentSubject);
                    activity.startActivity(intent);
                }catch (final JSONException e) {
                    //Log.e("COMAPANY JSON EXCEPTION", "Json parsing error: " + e.getMessage());
                    if (activity.getActivity() != null)
                        activity.getActivity().runOnUiThread(() -> Toast.makeText(activity.getContext(), "Json parsing error: " + e.getMessage(), Toast.LENGTH_LONG).show());
                }
            } else {
                if (activity.getActivity() != null)
                activity.getActivity().runOnUiThread(() -> Toast.makeText(activity.getContext(), "Couldn't get json from server. Check LogCat for possible errors!",
                        Toast.LENGTH_LONG)
                        .show());

            }
            return null;
        }

        @Override
        protected void onPostExecute(Void result) {
            super.onPostExecute(result);

        }
    }

    private static class JSONTransmitter extends AsyncTask<JSONObject, JSONObject, JSONObject> {
        private final WeakReference<HomeFragment> activityWeakReference;

        JSONTransmitter(HomeFragment context){
            activityWeakReference = new WeakReference<>(context);
        }

        @Override
        protected void onPreExecute() {
            HomeFragment activity = activityWeakReference.get();
            activity.pDialog = new ProgressDialog(activity.getContext());
            activity.pDialog.setMessage("Please wait...");
            activity.pDialog.setCancelable(false);
            activity.pDialog.show();
        }

        @Override
        protected JSONObject doInBackground(JSONObject... data) {
            final HomeFragment activity = activityWeakReference.get();
            JSONObject json = data[0];
            HttpClient client = new DefaultHttpClient();
            HttpConnectionParams.setConnectionTimeout(client.getParams(), 100000);
            JSONObject jsonResponse;
            String SIGNUP_URL = AppConfig.URL_REQUEST_BOOK;
            HttpPost post = new HttpPost(SIGNUP_URL);
            try {
                StringEntity se = new StringEntity( json.toString());
                post.addHeader("content-type", "application/json");
                post.addHeader("accept", "application/json");
                post.setEntity(se);
                HttpResponse response;
                response = client.execute(post);
                String resFromServer = EntityUtils.toString(response.getEntity());
                jsonResponse = new JSONObject(resFromServer);

                if (!jsonResponse.has("success")){
                    String success = jsonResponse.getString("success");
                    if (activity.getActivity() != null){
                        activity.getActivity().runOnUiThread(() -> Toast.makeText(activity.getContext(), success, Toast.LENGTH_SHORT).show());
                    }
                }else{
                    final String msg = jsonResponse.getString("error");
                    if (activity.getActivity() != null) {
                        activity.getActivity().runOnUiThread(() -> Toast.makeText(activity.getContext(), msg, Toast.LENGTH_SHORT).show());
                    }
                }
            } catch (Exception e) { e.printStackTrace();}

            return null;
        }

        @Override
        protected void onPostExecute(JSONObject jsonObject) {
            HomeFragment activity = activityWeakReference.get();
            if (activity.pDialog.isShowing()) {
                activity.pDialog.dismiss();
            }
        }
    }

    private void getUpdates(){
        db.collection("updates").get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
            @Override
            public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                List<Update> updates = queryDocumentSnapshots.toObjects(Update.class);
                updateList = new ArrayList<>();
                updateList.addAll(updates);
                SliderAdapter sliderAdapter = new SliderAdapter(getContext(), updateList);
                mViewPager.setAdapter(sliderAdapter);
                indicator.setupWithViewPager(mViewPager, true);
                Timer timer = new Timer();
                timer.scheduleAtFixedRate(new SliderTimer(), 4000, 6000);
            }
        });

    }
}

File: app/src/main/java/in/completecourse/questionbank/fragment/HomeFragment.kt
package `in`.completecourse.questionbank.fragment

import `in`.completecourse.questionbank.R
import `in`.completecourse.questionbank.ScanActivity
import `in`.completecourse.questionbank.SearchActivity
import `in`.completecourse.questionbank.SubjectActivity
import `in`.completecourse.questionbank.adapter.ImageAdapter
import `in`.completecourse.questionbank.adapter.SliderAdapter
import `in`.completecourse.questionbank.app.AppConfig
import `in`.completecourse.questionbank.helper.HelperMethods
import `in`.completecourse.questionbank.helper.HttpHandler
import `in`.completecourse.questionbank.helper.PrefManager
import `in`.completecourse.questionbank.model.CardModel
import `in`.completecourse.questionbank.model.Update
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.os.AsyncTask
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.ButterKnife
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import cz.msebera.android.httpclient.HttpResponse
import cz.msebera.android.httpclient.client.HttpClient
import cz.msebera.android.httpclient.client.methods.HttpPost
import cz.msebera.android.httpclient.entity.StringEntity
import cz.msebera.android.httpclient.impl.client.DefaultHttpClient
import cz.msebera.android.httpclient.params.HttpConnectionParams
import cz.msebera.android.httpclient.util.EntityUtils
import kotlinx.android.synthetic.main.fragment_home.*
import org.json.JSONException
import org.json.JSONObject
import java.lang.ref.WeakReference
import java.util.*

class HomeFragment : Fragment(), HelperMethods.ClickListener {
    private var updateList: ArrayList<Update>? = null
    private var pDialog: ProgressDialog? = null

    private var prefManager: PrefManager? = null

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == 100 && resultCode == 100 && data != null) {
            urlQR = data.getStringExtra("url")
            GetBookName(this@HomeFragment).execute()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val view: View = inflater.inflate(R.layout.fragment_home, container, false)
        ButterKnife.bind(this, view)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        prefManager = PrefManager(view.context.applicationContext)
        val cardList: ArrayList<CardModel> = ArrayList<CardModel>()
        cardList.add(CardModel(R.drawable.manual_search, "Manual Search"))
        cardList.add(CardModel(R.drawable.scan_qr, "Scan QR Code"))
        //use a linear layout manager
        val gridLayoutManager = GridLayoutManager(context, 2, RecyclerView.VERTICAL, false)
        recycler_view.layoutManager = gridLayoutManager

        //specify an adapter
        val recyclerViewAdapter = ImageAdapter(cardList, context)
        recycler_view.adapter = recyclerViewAdapter

        recycler_view.addOnItemTouchListener(HelperMethods.RecyclerTouchListener(context, this))

        if (isNetworkAvailable) {
            updates
        } else {
            Toast.makeText(view.context, "Please check your internet connection.", Toast.LENGTH_SHORT).show()
        }

        btn_request_now.setOnClickListener{
            requestBook()
        }

        //MediationTestSuite.launch(getContext());
    }

    fun requestBook() {
        if (isNetworkAvailable) {
            val id: String? = HelperMethods.generateChecksum()
            val userRequest = edTv_request_book.text.toString().trim { it <= ' ' }
            val mUserId: String? = prefManager?.userId
            if (userRequest.isNotEmpty()) {
                val dataObj = JSONObject()
                try {
                    dataObj.putOpt("id", id)
                    dataObj.putOpt("uid", mUserId)
                    dataObj.putOpt("userrequest", userRequest)
                    val jsonTransmitter = JSONTransmitter(this@HomeFragment)
                    jsonTransmitter.execute(dataObj)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            } else Toast.makeText(context, "Please enter your request", Toast.LENGTH_SHORT).show()
        } else Toast.makeText(context, "Please check your internet connection", Toast.LENGTH_SHORT).show()
    }

    private inner class SliderTimer : TimerTask() {
        override fun run() {
            if (activity != null) {
                activity!!.runOnUiThread {
                    if (viewPager.currentItem < updateList!!.size - 1) {
                        viewPager.currentItem = viewPager.currentItem + 1
                    } else {
                        viewPager.currentItem = 0
                    }
                }
            }
        }
    }

    /**
     * Checks if there is Internet accessible.
     * @return True if there is Internet. False if not.
     */
    private val isNetworkAvailable: Boolean get() {
            var activeNetworkInfo: NetworkInfo? = null
            if (activity != null) {
                val connectivityManager = activity!!.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
                activeNetworkInfo = connectivityManager.activeNetworkInfo
            }
            return activeNetworkInfo != null && activeNetworkInfo.isConnected
        }

    private class GetBookName(context: HomeFragment) : AsyncTask<Void?, Void?, Void?>() {
        private val activityWeakReference: WeakReference<HomeFragment> = WeakReference(context)

        override fun doInBackground(vararg arg0: Void?): Void? {
            val activity = activityWeakReference.get()
            val sh = HttpHandler()
            //tokenNumber = url.substring(url.lastIndexOf('/') + 1, url.length() - 4);
            //Log.e("token", tokenNumber);
            val url = "$urlQR/get"
            // Making a request to url and getting response
            val jsonStr = sh.makeServiceCall(url)
            //Log.e("Company Activity ",  "Response from url: " + jsonStr);
            if (jsonStr != null) {
                try {
                    val jsonObject = JSONObject(jsonStr)
                    val studentSubject = jsonObject.getString("subjectkiid")
                    val studentClass = jsonObject.getString("uclass")
                    //Log.e("class", studentClass);
                    //Log.e("subject", studentSubject);
                    val intent = Intent(activity!!.activity, SubjectActivity::class.java)
                    intent.putExtra("uclass", studentClass)
                    intent.putExtra("subjectkiid", studentSubject)
                    activity.startActivity(intent)
                } catch (e: JSONException) {
                    //Log.e("COMAPANY JSON EXCEPTION", "Json parsing error: " + e.getMessage());
                    if (activity!!.activity != null) activity.activity!!.runOnUiThread { Toast.makeText(activity.context, "Json parsing error: " + e.message, Toast.LENGTH_LONG).show() }
                }
            } else {
                if (activity!!.activity != null) activity.activity!!.runOnUiThread {
                    Toast.makeText(activity.context, "Couldn't get json from server. Check LogCat for possible errors!",
                            Toast.LENGTH_LONG)
                            .show()
                }
            }
            return null
        }

    }

    private class JSONTransmitter(context: HomeFragment) : AsyncTask<JSONObject?, JSONObject?, JSONObject?>() {
        private val activityWeakReference: WeakReference<HomeFragment> = WeakReference(context)

        override fun onPreExecute() {
            val activity = activityWeakReference.get()
            activity!!.pDialog = ProgressDialog(activity.context)
            activity.pDialog!!.setMessage("Please wait...")
            activity.pDialog!!.setCancelable(false)
            activity.pDialog!!.show()
        }

        override fun doInBackground(vararg data: JSONObject?): JSONObject? {
            val activity = activityWeakReference.get()
            val json = data[0]
            val client: HttpClient = DefaultHttpClient()
            HttpConnectionParams.setConnectionTimeout(client.params, 100000)
            val jsonResponse: JSONObject
            val SIGNUP_URL: String = AppConfig.URL_REQUEST_BOOK
            val post = HttpPost(SIGNUP_URL)
            try {
                val se = StringEntity(json.toString())
                post.addHeader("content-type", "application/json")
                post.addHeader("accept", "application/json")
                post.entity = se
                val response: HttpResponse
                response = client.execute(post)
                val resFromServer = EntityUtils.toString(response.entity)
                jsonResponse = JSONObject(resFromServer)
                if (!jsonResponse.has("success")) {
                    val success = jsonResponse.getString("success")
                    if (activity!!.activity != null) {
                        activity.activity!!.runOnUiThread { Toast.makeText(activity.context, success, Toast.LENGTH_SHORT).show() }
                    }
                } else {
                    val msg = jsonResponse.getString("error")
                    if (activity!!.activity != null) {
                        activity.activity!!.runOnUiThread { Toast.makeText(activity.context, msg, Toast.LENGTH_SHORT).show() }
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return null
        }

        override fun onPostExecute(jsonObject: JSONObject?) {
            val activity = activityWeakReference.get()
            if (activity!!.pDialog!!.isShowing) {
                activity.pDialog!!.dismiss()
            }
        }

    }

    private val updates: Unit get() {
            FirebaseFirestore.getInstance().collection("updates").get()
                    .addOnSuccessListener { queryDocumentSnapshots: QuerySnapshot ->
                        val updates = queryDocumentSnapshots.toObjects(Update::class.java)
                        updateList = ArrayList()
                        updateList!!.addAll(updates)
                        val sliderAdapter = SliderAdapter(context, updateList!!)
                        viewPager.adapter = sliderAdapter
                        indicator!!.setupWithViewPager(viewPager, true)
                        val timer = Timer()
                        timer.scheduleAtFixedRate(SliderTimer(), 4000, 6000)
                    }
        }

    companion object {
        private var urlQR: String? = null
    }

    override fun onClick(position: Int) {
        if (position == 0) {
            val intent = Intent(context, SearchActivity::class.java)
            startActivity(intent)
        } else if (position == 1) {
            val qrCodeActivityIntent = Intent(context, ScanActivity::class.java)
            startActivityForResult(qrCodeActivityIntent, 100)
        }
    }

}